### PR TITLE
Add Azure.Monitor.Query

### PR DESCRIPTION
**Pull request by pakrym**
Very first PR to get started with the service. No docs, minimal API surface, live test.

https://apiview.dev/Assemblies/Review/c2b04cb6ac7f475ab43e5f37e0226244

### CODE REVIEW

File: sdk/monitor/Azure.Monitory.Query/src/LogQueryRequest.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Monitory.Query.Models
```
Comment (christothes) [inline sdk/monitor/Azure.Monitory.Query/src/LogQueryRequest.cs:4]: is this intended to be a rename?

File: sdk/monitor/Azure.Monitory.Query/src/LogsBatchQuery.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            _clientDiagnostics = clientDiagnostics;
+            _restClient = restClient;
+            _batch = new BatchRequest();
+        }
+
+        protected LogsBatchQuery()
+        {
+        }
+
+        public virtual string Query(string workspaceId, string query)
```
Comment (christothes) [inline sdk/monitor/Azure.Monitory.Query/src/LogsBatchQuery.cs:31]: AddQuery?

File: sdk/monitor/Azure.Monitory.Query/src/LogsBatchQuery.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            _batch.Requests.Add(new LogQueryRequest()
+            {
+                Id = id,
+                Body = new QueryBody(query),
+                Workspace = workspaceId
+            });
+            return id;
+        }
+
+        public virtual Response<LogsBatchQueryResult> Execute(CancellationToken cancellationToken = default)
```
Comment (christothes) [inline sdk/monitor/Azure.Monitory.Query/src/LogsBatchQuery.cs:44]: Batches in other clients tend to use `Submit` here.

File: sdk/monitor/Azure.Monitory.Query/src/LogsBatchQueryResult.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    [CodeGenModel("BatchResponse")]
+    public partial class LogsBatchQueryResult
+    {
+        internal IReadOnlyList<LogQueryResponse> Responses { get; }
+        internal BatchResponseError Error { get; }
+
+        public LogsQueryResult GetResult(string queryId)
+        {
+            // TODO check status, add error message
+            return Responses.Single(r => r.Id == queryId).Body;
```
Comment (christothes) [inline sdk/monitor/Azure.Monitory.Query/src/LogsBatchQueryResult.cs:19]: Is throwing the desired behavior when there is no result for the given queryId?

File: sdk/monitor/Azure.Monitory.Query/src/MetricsClient.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            _metricDefinitionsClient = new MetricDefinitionsRestClient(_clientDiagnostics, _pipeline);
+            _metricsRestClient = new MetricsRestClient(_clientDiagnostics, _pipeline);
+            _namespacesRestClient = new MetricNamespacesRestClient(_clientDiagnostics, _pipeline);
+        }
+
+        protected MetricsClient()
+        {
+        }
+
+        public virtual Response<MetricQueryResult> Query(string resource, DateTimeOffset startTime, DateTimeOffset endTime, TimeSpan interval, CancellationToken cancellationToken = default)
```
Comment (christothes) [inline sdk/monitor/Azure.Monitory.Query/src/MetricsClient.cs:43]: arg check for `resource`?

File: sdk/monitor/Azure.Monitory.Query/src/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+// Replace with test project/test project public key and uncomment to make internal members visible to
```
Comment (christothes) [inline sdk/monitor/Azure.Monitory.Query/src/Properties/AssemblyInfo.cs:7]: replace?

File: sdk/monitor/Azure.Monitory.Query/tests/LogsQueryClientClientLiveTests.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+        [Test]
+        public async Task CanQuery()
+        {
+            var client = CreateClient();
+
+            var results = await client.QueryAsync(TestEnvironment.WorkspaceId, "Heartbeat");
+
+            var resultTable = results.Value.Tables.Single();
+            CollectionAssert.IsNotEmpty(resultTable.Columns);
```
Comment (christothes) [inline sdk/monitor/Azure.Monitory.Query/tests/LogsQueryClientClientLiveTests.cs:36]: validate that we returned the expected record?

File: sdk/monitor/Azure.Monitory.Query/tests/MetricsQueryClientLiveTests.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            var client = CreateClient();
+
+            var results = await client.QueryAsync(
+                TestEnvironment.MetricsResource,
+                Recording.Now.AddHours(-1),
+                Recording.Now,
+                TimeSpan.FromMinutes(1));
+
+            CollectionAssert.IsNotEmpty(results.Value.Metrics);
```
Comment (christothes) [inline sdk/monitor/Azure.Monitory.Query/tests/MetricsQueryClientLiveTests.cs:47]: validate that the time range was honored?

**Review by christothes — APPROVED** [review]

File: sdk/monitor/Azure.Monitory.Query/src/MetricsClient.cs [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        public MetricsClient(TokenCredential credential, MonitorQueryClientOptions options)
+        {
+            Argument.AssertNotNull(credential, nameof(credential));
+
+            options ??= new MonitorQueryClientOptions();
+
+            _clientDiagnostics = new ClientDiagnostics(options);
+            _pipeline = HttpPipelineBuilder.Build(options, new BearerTokenAuthenticationPolicy(credential, "https://management.azure.com//.default"));
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/monitor/Azure.Monitory.Query/src/MetricsClient.cs:33]: Yeah, half of this service is data plan over management entpoints.

File: sdk/monitor/Azure.Monitory.Query/src/LogQueryRequest.cs (hunk 1 above)
Comment (pakrym) [reply sdk/monitor/Azure.Monitory.Query/src/LogQueryRequest.cs:4]: Making the type internal.

File: sdk/monitor/Azure.Monitory.Query/src/LogsBatchQueryResult.cs (hunk 4 above)
Comment (pakrym) [reply sdk/monitor/Azure.Monitory.Query/src/LogsBatchQueryResult.cs:19]: Yeah, results for all queries have to be returned so yeah. But the exception should be way nicer.

File: sdk/monitor/Azure.Monitory.Query/src/MetricsClient.cs (hunk 5 above)
Comment (pakrym) [reply sdk/monitor/Azure.Monitory.Query/src/MetricsClient.cs:43]: Request building code would do it.

File: sdk/monitor/Azure.Monitory.Query/tests/LogsQueryClientClientLiveTests.cs (hunk 7 above)
Comment (pakrym) [reply sdk/monitor/Azure.Monitory.Query/tests/LogsQueryClientClientLiveTests.cs:36]: Unfortunately there is no data in that table. I'm working on how to get the test data into the service to write better tests.

File: sdk/monitor/Azure.Monitory.Query/api/Azure.Monitory.Query.netstandard2.0.cs [hunk 10]
```diff
@@ -0,0 +1,444 @@
+namespace Azure.Monitory.Query
+{
+    public partial class LogsClient
+    {
+        protected LogsClient() { }
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/monitor/Azure.Monitory.Query/api/Azure.Monitory.Query.netstandard2.0.cs:12]: I do, I thought I cleaned up the naming. 

https://apiview.dev/Assemblies/Review/c2b04cb6ac7f475ab43e5f37e0226244

Metric is the only single word name that's left.

File: sdk/monitor/Azure.Monitory.Query/src/LogsBatchQuery.cs (hunk 2 above)
Comment (pakrym) [reply sdk/monitor/Azure.Monitory.Query/src/LogsBatchQuery.cs:31]: Good one!

File: sdk/monitor/Azure.Monitory.Query/tests/MetricsQueryClientLiveTests.cs (hunk 8 above)
Comment (pakrym) [reply sdk/monitor/Azure.Monitory.Query/tests/MetricsQueryClientLiveTests.cs:47]: The data is not stable enough, working on finding ways to get it to be repeatable.

**Pull request merged**